### PR TITLE
feat(mcp): add session spawning helper

### LINKED ISSUE #291: [Feature]: Add MCP spawn_session interface
## Status Quo

Architect creates terminal sessions from in-app actions such as adding a terminal, opening a recent folder, switching worktrees, or restoring persisted sessions. Session creation is handled in `app/runtime.zig` through `SessionState.ensureSpawnedWithDir()`.

External integration is currently notification-only. Tools running inside Architect terminals can send status and story messages through `session/notify.zig`, but there is no MCP server and no external control interface that lets another local process ask Architect to create a terminal session.

## Objectives

Expose a small MCP interface that lets an external MCP client request a new Architect terminal session. The first supported operation is `spawn_session`, which starts a terminal in a requested working directory and can optionally run a command. This gives external tools a documented way to create sessions without UI automation.

## User Flow

**Trigger:** An MCP client calls `spawn_session`.

1. The client sends a `spawn_session` tool call with `cwd`, optional display metadata, and optional command details.
2. Architect validates the request and forwards it to the running app instance.
3. The app creates or selects an available terminal slot, spawns the session in `cwd`, and queues the command if one was provided.

**Result:** Architect displays a new terminal session and the MCP call returns a structured result.

## Scope

**In scope:**

- Add a local stdio MCP server executable named `architect-mcp`.
- Expose one MCP tool: `spawn_session`.
- Support a documented JSON input schema for `spawn_session`.
- Forward requests to the running Architect process through a local control channel.
- Return a structured result with status and a session identifier when available.
- Return clear errors for invalid input, missing running app, full grid, invalid `cwd`, and spawn failures.

**Out of scope:**

- Any domain-specific workflow built on top of MCP.
- Session monitoring after spawn.
- Sending input into existing sessions.
- Approval handling or artifact UI.
- Dynamic agent routing or capability detection.
- Remote HTTP MCP transport.

## Deployment

`architect-mcp` should be a separate helper executable, not a special mode of the main `architect` GUI executable.

- Source builds install both executables under `zig-out/bin/`: `architect` and `architect-mcp`.
- macOS release bundles include the helper at `Architect.app/Contents/MacOS/architect-mcp`. The app entrypoint remains `Architect.app/Contents/MacOS/architect`.
- The bundle script signs `architect-mcp` along with the main executable and bundled libraries.
- Homebrew installs the helper inside `Architect.app/Contents/MacOS/` and should expose a convenient `architect-mcp` command on `PATH` when practical. Documentation must also show the app-bundle path as the fallback command.
- `architect-mcp` is a stdio MCP server process launched by an MCP client. It is not a daemon. Each invocation connects to the running Architect app through the local control channel.
- If Architect is not running, `architect-mcp` returns a clear MCP tool error. It should not launch the GUI app itself.

### Affected Modules

- `src/mcp/` or `src/mcp_main.zig`: MCP stdio server executable and tool schema.
- `src/app/runtime.zig`: request draining and session-spawn handling.
- `src/session/` or `src/app/`: local control socket and request queue.
- `build.zig`: build and install `architect-mcp` next to `architect`.
- `scripts/bundle-macos.sh`: copy and sign `architect-mcp` inside `Architect.app/Contents/MacOS/`.
- `.github/workflows/release.yaml`: pass the helper executable into the bundle step or otherwise ensure it is included in release archives.
- `Formula/architect.rb`: install the helper in Homebrew packages and expose a convenient command when practical.
- `README.md`, `docs/ai-integration.md`, `docs/ARCHITECTURE.md`: document the interface, deployment paths, and architecture.

### Tasks

Each task is independently testable. Ordered by implementation sequence.

1. [ ] Define the `spawn_session` input and output schema.
2. [ ] Add a local control channel for external spawn requests.
3. [ ] Implement runtime handling for validated spawn requests.
4. [ ] Implement the stdio MCP server lifecycle, `tools/list`, and `tools/call`.
5. [ ] Wire the new executable into `build.zig`.
6. [ ] Update macOS release bundling so `architect-mcp` is included in `Architect.app/Contents/MacOS/` and signed.
7. [ ] Update the Homebrew formula to install the helper and, when practical, expose `architect-mcp` on `PATH`.
8. [ ] Write tests for schema validation, MCP responses, control-channel errors, and spawn-request handling.
9. [ ] Update `README.md`.
10. [ ] Update `docs/ai-integration.md`.
11. [ ] Update `docs/ARCHITECTURE.md`.

### New Dependencies

None. Use Zig stdlib for the minimal stdio JSON-RPC/MCP implementation.

## Acceptance Criteria

- [ ] `architect-mcp` starts as a stdio MCP server.
- [ ] Source builds install `zig-out/bin/architect-mcp`.
- [ ] macOS release archives include `Architect.app/Contents/MacOS/architect-mcp`.
- [ ] The helper is signed as part of the app bundle.
- [ ] `tools/list` exposes exactly one tool: `spawn_session`.
- [ ] `spawn_session` validates input and returns structured success or error results.
- [ ] A successful call creates a new Architect terminal session in the requested working directory.
- [ ] Optional command details are safely passed to the spawned terminal.
- [ ] Failure cases do not crash Architect or the MCP server.
- [ ] Tests cover the happy path and documented edge cases.
- [ ] `zig build`, `zig build test`, and `just lint` pass.
- [ ] Documentation is updated.

### PR DESCRIPTION
**Pull request by forketyfork**
## Solution

Adds a separate `architect-mcp` stdio helper for local MCP clients. It exposes `spawn_session`, validates the tool arguments, and forwards requests to the running Architect app through a Unix-domain control socket discovered from the runtime temp directory.

The app handles those requests on the main loop: it validates the working directory, picks or expands a grid slot, spawns the terminal session, queues optional command text, focuses the new session, and returns structured success or stable error codes. Build, release packaging, Homebrew, and docs now ship and describe the helper.

Closes #291

## Checks

- `nix develop -c zig build`
- `nix develop -c zig build test`
- `nix develop -c just lint`
- `git diff --check`
- MCP smoke checks for `tools/list` and `spawn_session`

## Test plan

- [ ] Start Architect, call `spawn_session` through `architect-mcp` with an existing absolute `cwd`, and confirm a new focused session appears.
- [ ] Call `spawn_session` with a `command` value and confirm the command is submitted in the new shell.

### CODE REVIEW

File: src/app/control.zig [hunk 1]
```diff
@@ -0,0 +1,736 @@
...
+    }
+}
+
+fn handleControlConnection(
+    allocator: std.mem.Allocator,
+    conn_fd: posix.fd_t,
+    queue: *SpawnQueue,
+    runtime_wake: ?RuntimeWake,
+) void {
+    const bytes = readLineFromFd(allocator, conn_fd, max_message_bytes) catch |err| {
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/app/control.zig]: Fixed in `6bdd5a2`. Accepted control connections are now set non-blocking, and request reads have a bounded deadline, so a partial local client cannot pin the single listener thread indefinitely.

File: src/app/control.zig [hunk 2]
```diff
@@ -0,0 +1,736 @@
...
+    };
+}
+
+fn writeDiscoveryFile(allocator: std.mem.Allocator, path: []const u8, socket_path: []const u8) !void {
+    const payload = try discoveryPayloadAlloc(allocator, socket_path);
+    defer allocator.free(payload);
+
+    const file = try std.fs.createFileAbsolute(path, .{ .truncate = true });
+    defer file.close();
+    try file.writeAll(payload);
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/app/control.zig]: Fixed in `6bdd5a2`. The discovery JSON is written to a temporary file with `0600` permissions, synced, and then renamed into place. The app also logs the full discovery JSON path and socket path after publishing it.

File: src/app/control.zig [hunk 3]
```diff
@@ -0,0 +1,736 @@
+const std = @import("std");
+const posix = std.posix;
+const atomic = std.atomic;
+
+const log = std.log.scoped(.control);
+
+pub const max_message_bytes: usize = 64 * 1024;
+const max_cwd_bytes: usize = 4096;
+const max_command_bytes: usize = 16 * 1024;
+const discovery_file_name = "architect_control.json";
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/app/control.zig]: Fixed in `6bdd5a2`. Discovery files are now per user and process, and `architect-mcp` scans for the newest reachable instance instead of relying on one shared `architect_control.json`.

File: src/app/control.zig (hunk 1 above)
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/app/control.zig]: Fixed in `6bdd5a2`. The accepted fd is now non-blocking, and the control request read path times out if a complete line does not arrive.

File: src/app/control.zig [hunk 4]
```diff
@@ -0,0 +1,736 @@
...
+
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/app/control.zig]: Fixed in `6bdd5a2`. The discovery filename now includes uid and pid, and the client filters discovery files for the current user before trying reachable sockets.

File: src/mcp/main.zig [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+                    buffer.clearRetainingCapacity();
+                }
+                continue;
+            }
+            if (byte == '\r') continue;
+            if (buffer.items.len >= control.max_message_bytes) {
+                try writeJsonRpcError(allocator, stdout_file, null, .invalid_request, "message is too large");
+                buffer.clearRetainingCapacity();
+                continue;
+            }
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/mcp/main.zig:55]: Fixed in `6bdd5a2`. After reporting an oversized stdin message, `architect-mcp` now discards bytes until the next newline before parsing again.
